lxd/device/tpm.go

```diff
@@ -0,0 +1,262 @@
+package device
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/revert"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/subprocess"
+	"github.com/lxc/lxd/shared/validate"
+)
+
+type tpm struct {
+	deviceCommon
+}
+
+// isRequired indicates whether the device config requires this device to start OK.
+func (d *tpm) isRequired() bool {
+	// Defaults to not required.
+	if shared.IsTrue(d.config["required"]) {
+		return true
+	}
+
+	return false
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *tpm) validateConfig(instConf instance.ConfigReader) error {
+	if !instanceSupported(instConf.Type(), instancetype.Container, instancetype.VM) {
+		return ErrUnsupportedDevType
+	}
+
+	rules := map[string]func(string) error{
+		"required": validate.Optional(validate.IsBool),
+	}
+
+	err := d.config.Validate(rules)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateEnvironment checks if the TPM emulator is available.
+func (d *tpm) validateEnvironment() error {
+	// Validate the required binary.
+	_, err := exec.LookPath("swtpm")
+	if err != nil {
+		return fmt.Errorf("Required tool '%s' is missing", "swtpm")
+	}
+
+	if d.inst.Type() == instancetype.Container {
+		// Load module tpm_vtpm_proxy which creates the /dev/vtpmx device, required
+		// by the TPM emulator.
+		module := "tpm_vtpm_proxy"
+
+		err := util.LoadModule(module)
+		if err != nil {
+			return fmt.Errorf("Failed to load kernel module %q: %s", module, err)
+		}
+	}
+
+	return nil
+}
+
+// Start is run when the device is added to the instance.
+func (d *tpm) Start() (*deviceConfig.RunConfig, error) {
+	err := d.validateEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	if d.inst.Type() == instancetype.VM {
+		return d.startVM()
+	}
+
+	return d.startContainer()
+}
+
+func (d *tpm) startContainer() (*deviceConfig.RunConfig, error) {
+	tpmDevPath := filepath.Join(d.inst.Path(), fmt.Sprintf("tpm-%s", d.name))
+	logFileName := fmt.Sprintf("tpm.%s.log", d.name)
+	logPath := filepath.Join(d.inst.LogPath(), logFileName)
+
+	if !shared.PathExists(tpmDevPath) {
+		err := os.Mkdir(tpmDevPath, 0755)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	proc, err := subprocess.NewProcess("swtpm", []string{"chardev", "--tpm2", "--tpmstate", fmt.Sprintf("dir=%s", tpmDevPath), "--vtpm-proxy"}, logPath, "")
```

Comment:
You don't appear to be saving a pidfile for the container case, how is the daemon getting stopped in this case?

Reply:
I forgot to add that. Fixed now.